src/circuit/gadgets/constraints.rs

```diff
@@ -430,6 +503,443 @@ pub fn must_be_simple_bit(x: &Boolean) -> AllocatedBit {
+    /*let p64_bn = match g.power2_64_num.get_value() {
+        Some(v) => BigUint::from_bytes_le(v.to_repr().as_ref()),
+        // Since it is will be in the denominator, it can't be zero
+        None => BigUint::one(), // blank and dummy
+    };*/
+    let p64_bn = BigUint::pow(&BigUint::from_u32(2).unwrap(), 64);
+    let v = match maybe_u64.hash().get_value() {
+        Some(v) => v,
+        None => F::zero(),
+    };
+    let u64_ptr = store.get_u64(v.to_u64_unchecked());
+    let field_bn = BigUint::from_bytes_le(v.to_repr().as_ref());
+
+    let (q_bn, r_bn) = field_bn.div_rem(&p64_bn);
+
+    let q_num = allocate_unconstrained_bignum(&mut cs.namespace(|| "q"), q_bn)?;
+    let r_num = allocate_unconstrained_bignum(&mut cs.namespace(|| "r"), r_bn)?;
+    //let p64_num = allocate_unconstrained_bignum(&mut cs.namespace(|| "p64"), p64_bn)?;
+    //let p64_num = g.power2_64_num;
+
+    // a = b.q + r
+    let product = mul(
+        &mut cs.namespace(|| "product(q,pow(2,64))"),
+        &q_num,
+        &g.power2_64_num,
+    )?;
+    let sum = add(&mut cs.namespace(|| "sum remainder"), &product, &r_num)?;
+    let u64_decomp = alloc_equal(
+        &mut cs.namespace(|| "check u64 decomposition"),
+        &sum,
+        maybe_u64.hash(),
+    )?;
+    enforce_true(
+        &mut cs.namespace(|| "enforce u64 decomposition"),
+        &u64_decomp,
+    )?;
+
+    // enforce remainder range
+    enforce_at_most_n_bits(
+        &mut cs.namespace(|| "remainder u64 range"),
+        g,
+        r_num.clone(),
+        64,
+    )?;
+
+    let output = AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "u64_op"), store, || Ok(&u64_ptr))?;
+    let u64_is_remainder = alloc_equal(
+        &mut cs.namespace(|| "check u64 value is the remainder"),
+        &r_num,
+        output.hash(),
+    )?;
+    enforce_true(
+        &mut cs.namespace(|| "enforce u64 decomposition output"),
+        &u64_is_remainder,
+    )?;
+    Ok(output)
+}
+
+pub fn hide<F: LurkField, CS: ConstraintSystem<F>>(
+    mut cs: CS,
+    g: &GlobalAllocations<F>,
+    secret: &AllocatedNum<F>,
+    maybe_payload: &AllocatedPtr<F>,
+    store: &Store<F>,
+) -> Result<AllocatedPtr<F>, SynthesisError> {
+    AllocatedPtr::construct_commitment(
+        &mut cs.namespace(|| "commitment hash"),
+        g,
+        store,
+        secret,
+        maybe_payload,
+    )
+}
+
+pub fn num<F: LurkField, CS: ConstraintSystem<F>>(
+    mut cs: CS,
+    maybe_num: &AllocatedPtr<F>,
+    store: &Store<F>,
+) -> Result<AllocatedPtr<F>, SynthesisError> {
+    let num_ptr = if let Some(ptr) = maybe_num.ptr(store).as_ref() {
+        let scalar_ptr = store.get_expr_hash(ptr).expect("expr hash missing");
+        match store.get_num(crate::Num::Scalar::<F>(*scalar_ptr.value())) {
+            Some(n) => n,
+            None => store.get_nil(),
+        }
+    } else {
+        store.get_nil()
+    };
+    AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "num"), store, || Ok(&num_ptr))
```

Comment:
Although this (along with `comm` and `char_op` below) were not introduced in this work, the fact of moving them into `constraints.rs` makes clear that they suffer a problem we need to address — either here or immediately otherwise. The problem is that the allocation is unconstrained.

It's also problematic to have the calls to `store.get_nil()` in `constraints.rs` — which makes me think we should not move these function here as they are.

Let's leave the functions where they are, for now. I'll create an issue to address the problems. If appropriate, any general-purpose helpers can be moved here during the process of fixing.

Reply:
Actually, the good news is that this was fixed in #203 (thought it was familiar). However, note that the fix would have been lost if this PR was merged, so please rebase such that it is preserved.